### ISSUE TITLE
Header parsing

### ISSUE THREAD

**Issue by samiam95124** [samiam95124/Pascal-P6#157]
# Adding header parsing

## The state now

The current state of Pascal-P6 is that the header files are parsed automatically, and any options are parsed manually. The drawbacks to this are:

1. The options must appear at the end, since the header files are already parsed before the program executes.
2. Filenames are parsed from the command line, even if the file is not going to be used (like prr and prd files).

The advantage of the current system is it stays completely compatible with ISO 7185 Pascal.

## Proposed

The proposal involves two steps:

1. An option is added that stops parsing of the header files.
2. The header file parsing is expanded to include file parsing and options anywhere on the command line.

The result will still be compatible with ISO 7185 by option. The options will be parsed no matter where they are on the command line. For files, one or more files can be elided. In this case, reading from or writing to those elided files yields an error.

**Comment by samiam95124** [comment]
This is done in two steps. The first is to parse the input files just the way that the automated method does, so that it is backward compatible. The second is to consider skipping files that have no use. For example, pint may or may not use either or both of its prd and prr files. Since this is optional, this should also be backwards compatible since the scripts always provide the filename, even if not used (a dummy).

Basically this brings the individual atomic commands to their final form, which is to handle command line files and options in a flexible manner.

**Comment by samiam95124** [comment]
Another thing this enables is to automatically fill in extensions like .pas, p6, .s etc. This could be done on this ticket or a follow on.

**Comment by samiam95124** [comment]
As far as making header files optional, will also need to create a flag for parsed/non-parsed on files that are elided. For example, pcom can elide its output file. There is actually an option for this, prcode, that stops the output to that file from happening. This option should be automatically set if the output file is elided. However, a flag like "prrelide" needs to exist that is set if the header file is elided. In that way, if the user attempts to turn the output back on in a source option statement, it should give an error like "output file not specified" or similar.

**Comment by samiam95124** [comment]
Did the output file optional thing on pcom.

**Comment by samiam95124** [comment]
Added default extension adding.

Now pcom accepts any of these forms:

pcom test.pas test.p6

pcom test.pas - Output file is supressed (compile only).

pcom test test - both .pas and .p6 extensions added automatically.

pcom test - .pas extension added and output file supressed.

Plus --options can appear anywhere in the command line, before the files, between them, or after.

This will get extended to pint and pmach before closing this ticket.

**Comment by samiam95124** [comment]
Notes on programs:

pint defaults to .p6 on input, and .out on output. If it is used to produce a binary machine deck, or .p6o file, that must be specified.

Eliding the output file on pint is valid, but produces an error if the program attempts to write to the prr file.

pmach is similar, but defaults to .p6o on input. Output is .out. As per pint, it faults if the output is elided and a write to prr occurs.

**Comment by samiam95124** [comment]
This is complete. I did't do the elide of the output file on other binaries besides pcom, because it is a fair amount of work. Perhaps later.
